ihmc-perception/src/main/java/us/ihmc/perception/gpuHeightMap/CUDAFootstepOptimizer.java

```diff
@@ -0,0 +1,321 @@
+                          .withPointer(gpuBestCosts)
+                          .withPointer(gpuBestIndices)
+                          .run(cudaStream, gridSize, blockSize, 0);
+
+         FloatPointer gpuGlobalBestCost = new FloatPointer();
+         IntPointer gpuGlobalBestIndex = new IntPointer();
+         CUDATools.mallocAsync(gpuGlobalBestIndex, 1, cudaStream);
+         CUDATools.mallocAsync(gpuGlobalBestCost, 1, cudaStream);
+
+         LogTools.warn("Running global result kernel");
+         // Run the kernel
+         blockSize.x(resultSize);
+         gridSize.x(1);
+         globalResultKernel.withPointer(gpuBestCosts)
+                           .withPointer(gpuBestIndices)
+                           .withInt(resultSize)
+                           .withPointer(gpuGlobalBestCost)
+                           .withPointer(gpuGlobalBestIndex)
+                           .run(cudaStream, gridSize, blockSize, 0);
+         LogTools.warn("Done global result kernel");
+         IntPointer cpuGlobalBestIndex = new IntPointer(1);
+         FloatPointer cpuGlobalBestCost = new FloatPointer(1);
+         CUDATools.memcpyAsync(cpuGlobalBestIndex, gpuGlobalBestIndex, 1, cudaStream);
+         CUDATools.memcpyAsync(cpuGlobalBestCost, gpuGlobalBestCost, 1, cudaStream);
+
+         cudart.cudaStreamSynchronize(cudaStream);
+
+         int startIndex = cpuGlobalBestIndex.get(0);
+         LogTools.info("Grid search finished. Best cost: {}, idx: {}", cpuGlobalBestCost.get(0), startIndex);
+         bestSolution[0] = cpuSolutions.get(startIndex * 3L);
+         bestSolution[1] = cpuSolutions.get(startIndex * 3L + 1);
+         bestSolution[2] = cpuSolutions.get(startIndex * 3L + 2);
+
+         FramePose3D optimalFootstepPose = new FramePose3D(initialPose);
+         optimalFootstepPose.setX(bestSolution[0]);
+         optimalFootstepPose.setY(bestSolution[1]);
+         optimalFootstepPose.getRotation().set(new YawPitchRoll(bestSolution[2], initialPose.getPitch(), initialPose.getRoll()));
+         optimalFootstepPose.setZ(currentHeightMapData.getHeightAt(bestSolution[0], bestSolution[1]));
+
+         // Release stuff
+         cudaFreeAsync(gpuInitialPose, cudaStream);
+         cudaFreeAsync(gpuHeights, cudaStream);
+         cudaFreeAsync(gpuGridCenter, cudaStream);
+         cudaFreeAsync(gpuCosts, cudaStream);
+         cudaFreeAsync(gpuGlobalBestCost, cudaStream);
+         cudaFreeAsync(gpuGlobalBestIndex, cudaStream);
+         cudaFreeAsync(gpuSolutions, cudaStream);
+         cudaFreeAsync(gpuBestCosts, cudaStream);
+         cudaFreeAsync(gpuBestIndices, cudaStream);
+
+         return optimalFootstepPose;
+      }
+   }
+
+   @Override
+   public void close()
+   {
+      blockSize.close();
+      gridSize.close();
+      computeKernel.close();
+      blockResultKernel.close();
+      globalResultKernel.close();
+      program.close();
+      CUDAStreamManager.releaseStream(cudaStream);
+   }
+
+   public void testResultKernel()
```

Comment:
Same here. Should be package-private
```suggestion
   void testResultKernel()
```